### ISSUE TITLE
qwen38flashnext cookbook image: 4 bugs on single-device SM121 (GB10 Spark) — silent garbage decode (trtllm-gen), non-compacting _compact_kv, TMA-O varlen boot crash, fp8 tl.dot in sparse prefill

### ISSUE THREAD

**Issue by Death-By-Tokens** [sgl-project/sglang#36716]
**Image:** `lmsysorg/sglang:qwen38flashnext` (the Qwen3.8-Flash-Next cookbook build)
**Hardware:** single DGX Spark (GB10, SM121), tp=1, `--quantization modelopt_fp4`, fp8 KV
**Context:** we run Flash-Next on one Spark (reference recipe is 2x Spark TP2) and hit four reproducible bugs on paths the TP2 recipe never exercises. Working fixes for all four are bind-mounted patch files in https://github.com/Death-By-Tokens/Qwen3.8-Flash-Next-180B-on-ONE-DGX-Spark (see `patches/` + `patches/generators/` for exact anchored diffs). Happy to turn any of these into PRs.

### 1. flashinfer trtllm-gen decode kernels silently emit garbage on SM121

`qwen_sparse_attn_backend.py` prefers `trtllm_batch_decode_with_kv_cache` whenever importable. trtllm-gen attention kernels are SM100-only (the image itself logs this for MoE and falls back), but the QSA decode resolver has no arch guard — on SM121 the kernel runs and **silently corrupts output**: first decoded token is correct, then permanent `!!!!` (NaN logits -> token 0). No error, no crash, `/health` stays green.

Suggested fix: gate `_resolve_trtllm_sparse_decode()` on SM100, same as the MoE runner selection already does.

### 2. `_compact_kv` does not compact — NaN holes on the FA fallback path

With trtllm decode disabled (or unavailable), decode uses the packed FA fallback. `qwen_sparse_kv_extraction_compact_triton` writes valid rows at their **original column offsets** (`dst = pack_start + cols`) while `cu_seqlens_k`/`valid_counts` promise a **contiguous** segment. Any interleaved `-1` in `topk_indices` therefore leaves uninitialized scratch rows *inside* the segment the attention then reads -> NaN -> same `!!!!` collapse. This path appears untested because SM100 deployments always take trtllm decode.

Our workaround replaces the pack+FA step with a direct `req_to_token` gather + masked SDPA (exact for q_len==1, hole-tolerant, CUDA-graph-safe); a proper fix is a true compaction in the kernel.

### 3. flash-attn cute `flash_fwd.py` (as shipped in the image): TMA-O enabled for varlen

The image's copy has:

```python
# self.use_tma_O = self.arch >= 90 and mCuSeqlensQ is None
self.use_tma_O = self.arch >= Arch.sm_90
```

i.e. the varlen guard is commented out. The ragged epilogue then indexes a rank-2 view with a rank-3 coordinate and every varlen compile dies at boot with MLIR `expects 'coord' and shape of view are weakly congruent`. Restoring `and mCuSeqlensQ is None` fixes boot. (Filed here rather than Dao-AILab because the modified copy ships inside this image; unclear if upstream is affected.)

### 4. Long-prefill sparse kernel: `tl.dot` on fp8-loaded K

With `--kv-cache-dtype fp8_e4m3`, `qsa/sparse_attn.py`'s sparse prefill kernel feeds K loaded straight from the fp8 cache into `tl.dot` -> `Unsupported rhs dtype fp8e4nv`. The kernel variant only compiles at long-prefill shapes, so short prompts work fine and the server dies (SIGQUIT) on the first ~100k-token request. Fix: `keys.to(q_values.dtype)` before the QK dot and compute the P*V dot in bf16 instead of `probabilities.to(values.dtype)`.

### Bonus observation (not a bug report): speculative decoding requires the mamba state-checkpoint flags

With NEXTN, omitting `--mamba-scheduler-strategy extra_buffer --mamba-track-interval 64` looks fine on prose and corrupts mid-stream on high-acceptance (code) output — rejected drafts need DeltaNet state rollback. Might be worth a startup warning when spec decode is enabled without state tracking.

---
All four fixes are validated end-to-end on the linked repo's build: 262,144-token context, exact needle recall at 222k tokens, 12/12 executed-code benchmark, 27-46 tok/s decode. Happy to submit PRs for any subset if useful.

**Comment by jahnclawdmonet** [comment]
Bug 3 provenance settled: it is not an image-local modification. The stock PyPI wheel ships this way.

`flash_attn_4-4.0.0b19-py3-none-any.whl` downloaded fresh from PyPI contains the same commented-out guard at `flash_attn/cute/flash_fwd.py:658-659`:

```python
# self.use_tma_O = self.arch >= 90 and mCuSeqlensQ is None
self.use_tma_O = self.arch >= Arch.sm_90
```

and the installed copy in `lmsysorg/sglang:qwen38flashnext` is byte-identical to the wheel's (md5 `ba105efb753685941dc798fa339b5d26` on both). The sglang-vendored copy in the same image (`sglang/kernels/ops/attention/flash_attn/cute/flash_fwd.py`) carries the same pattern at lines 872-873. So upstream flash-attn-4 is affected as released and this belongs in Dao-AILab's tracker as well; the MLIR `weakly congruent` boot crash tracked in #36531 / #36558 is this bug reached through the QSA varlen fallback.

On bug 1, a datapoint from the other side of the family split that matters for the fix shape: on sm_120 (RTX PRO 6000) the trtllm-gen QSA decode path does not corrupt. With PR #36556's gate applied and real Flash-Next-NVFP4 weights, QSA decode output is byte-identical to a `--decode-attention-backend flashinfer` override at 2,125 / 21,170 / 43,317 prompt tokens (A/B details in the PR thread), and two further sm_120 users report correct service on that path, one at full native context. So gating `_resolve_trtllm_sparse_decode()` to SM100 alone would trade sm_121 correctness for an sm_120 regression onto the fallback path your bug 2 shows is also broken. The data supports a per-arch gate: sm_100 and sm_120 in, sm_121 out, with your bug 2 fix making the fallback safe wherever it is taken.
